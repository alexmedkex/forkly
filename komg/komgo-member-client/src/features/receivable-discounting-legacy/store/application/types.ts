import { ImmutableMap, stringOrNull } from '../../../../utils/types'
import {
  IReceivablesDiscountingInfo,
  IReceivablesDiscountingBase,
  ICreateReceivableDiscountingApplicationResponse,
  IHistory,
  IReceivablesDiscounting
} from '@komgo/types'
import { List, Map } from 'immutable'
import { Action } from 'redux'

export enum ReceivableDiscountingApplicationActionType {
  // create
  CREATE_APPLICATION_REQUEST = '@@receivable-discounting/CREATE_APPLICATION_REQUEST',
  CREATE_APPLICATION_SUCCESS = '@@receivable-discounting/CREATE_APPLICATION_SUCCESS',
  CREATE_APPLICATION_FAILURE = '@@receivable-discounting/CREATE_APPLICATION_FAILURE',

  // update
  UPDATE_APPLICATION_REQUEST = '@@receivable-discounting/UPDATE_APPLICATION_REQUEST',
  UPDATE_APPLICATION_SUCCESS = '@@receivable-discounting/UPDATE_APPLICATION_SUCCESS',
  UPDATE_APPLICATION_FAILURE = '@@receivable-discounting/UPDATE_APPLICATION_FAILURE',

  SHARE_APPLICATION_REQUEST = '@@receivable-discounting/SHARE_APPLICATION_REQUEST',
  SHARE_APPLICATION_SUCCESS = '@@receivable-discounting/SHARE_APPLICATION_SUCCESS',
  SHARE_APPLICATION_FAILURE = '@@receivable-discounting/SHARE_APPLICATION_FAILURE',

  // read
  FETCH_APPLICATION_REQUEST = '@@receivable-discounting/FETCH_APPLICATION_REQUEST',
  FETCH_APPLICATION_SUCCESS = '@@receivable-discounting/FETCH_APPLICATION_SUCCESS',
  FETCH_APPLICATION_FAILURE = '@@receivable-discounting/FETCH_APPLICATION_FAILURE',

  FETCH_MULTIPLE_APPLICATION_REQUEST = '@@receivable-discounting/FETCH_MULTIPLE_APPLICATION_REQUEST',
  FETCH_MULTIPLE_APPLICATION_SUCCESS = '@@receivable-discounting/FETCH_MULTIPLE_APPLICATION_SUCCESS',
  FETCH_MULTIPLE_APPLICATION_FAILURE = '@@receivable-discounting/FETCH_MULTIPLE_APPLICATION_FAILURE',

  FETCH_APPLICATION_HISTORY_REQUEST = '@@receivable-discounting/FETCH_APPLICATION_HISTORY_REQUEST',
  FETCH_APPLICATION_HISTORY_SUCCESS = '@@receivable-discounting/FETCH_APPLICATION_HISTORY_SUCCESS',
  FETCH_APPLICATION_HISTORY_FAILURE = '@@receivable-discounting/FETCH_APPLICATION_HISTORY_FAILURE'
}

export interface ApplicationError extends Action {
  type:
    | ReceivableDiscountingApplicationActionType.CREATE_APPLICATION_FAILURE
    | ReceivableDiscountingApplicationActionType.FETCH_MULTIPLE_APPLICATION_FAILURE
  payload: stringOrNull
}

export interface CreateSucceeded extends Action {
  type: ReceivableDiscountingApplicationActionType.CREATE_APPLICATION_SUCCESS
  payload: ICreateReceivableDiscountingApplicationResponse
}

export interface FetchMultipleSucceeded extends Action {
  type: ReceivableDiscountingApplicationActionType.FETCH_MULTIPLE_APPLICATION_SUCCESS
  payload: IReceivablesDiscountingInfo[]
}

export interface FetchSucceeded extends Action {
  type: ReceivableDiscountingApplicationActionType.FETCH_APPLICATION_SUCCESS
  payload: IReceivablesDiscountingInfo
}

export interface UpdateSucceeded extends Action {
  type: ReceivableDiscountingApplicationActionType.UPDATE_APPLICATION_SUCCESS
  payload: IReceivablesDiscountingBase
  rdId: string
}

export interface FetchHistorySucceeded extends Action {
  type: ReceivableDiscountingApplicationActionType.FETCH_APPLICATION_HISTORY_SUCCESS
  payload: IHistory<IReceivablesDiscounting>
  rdId: string
}

export interface ReceivableDiscountingApplicationStateProperties {
  byId: Map<string, IReceivablesDiscountingInfo>
  historyById: Map<string, IHistory<IReceivablesDiscounting>>
  ids: List<string>
  error: stringOrNull
}

export type ReceivableDiscountingApplicationState = ImmutableMap<ReceivableDiscountingApplicationStateProperties>

export type ReceivableDiscountingApplicationAction =
  | CreateSucceeded
  | FetchHistorySucceeded
  | FetchMultipleSucceeded
  | FetchSucceeded
  | UpdateSucceeded
  | ApplicationError
