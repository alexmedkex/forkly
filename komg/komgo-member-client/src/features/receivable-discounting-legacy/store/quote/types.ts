import { IQuoteBase, IQuote, IHistory } from '@komgo/types'
import { Action } from 'redux'
import { stringOrNull, ImmutableMap } from '../../../../utils/types'
import { Map } from 'immutable'

export enum QuoteActionType {
  // create
  SHARE_QUOTE_REQUEST = '@@receivable-discounting/SHARE_QUOTE_REQUEST',
  SHARE_QUOTE_SUCCESS = '@@receivable-discounting/SHARE_QUOTE_SUCCESS',
  SHARE_QUOTE_FAILURE = '@@receivable-discounting/SHARE_QUOTE_FAILURE',

  // update
  UPDATE_QUOTE_REQUEST = '@@receivable-discounting/UPDATE_QUOTE_REQUEST',
  UPDATE_QUOTE_SUCCESS = '@@receivable-discounting/UPDATE_QUOTE_SUCCESS',
  UPDATE_QUOTE_FAILURE = '@@receivable-discounting/UPDATE_QUOTE_FAILURE',

  // read
  FETCH_QUOTE_REQUEST = 'FETCH_QUOTE_REQUEST',
  FETCH_QUOTE_FAILURE = 'FETCH_QUOTE_FAILURE',
  FETCH_QUOTE_SUCCESS = 'FETCH_QUOTE_SUCCESS',

  FETCH_QUOTE_HISTORY_REQUEST = 'FETCH_QUOTE_HISTORY_REQUEST',
  FETCH_QUOTE_HISTORY_SUCCESS = 'FETCH_QUOTE_HISTORY_SUCCESS',
  FETCH_QUOTE_HISTORY_FAILURE = 'FETCH_QUOTE_HISTORY_FAILURE'
}

export interface UpdateSucceeded extends Action {
  type: QuoteActionType.UPDATE_QUOTE_SUCCESS
  payload: IQuoteBase
}

export interface QuoteStateProperties {
  byId: Map<string, IQuote>
  historyById: Map<string, IHistory<IQuote>>
  error: stringOrNull
}

export interface FetchHistorySucceeded extends Action {
  type: QuoteActionType.FETCH_QUOTE_HISTORY_SUCCESS
  payload: IHistory<IQuote>
  quoteId: string
}

export type QuoteState = ImmutableMap<QuoteStateProperties>

export type QuoteAction = UpdateSucceeded | FetchHistorySucceeded
