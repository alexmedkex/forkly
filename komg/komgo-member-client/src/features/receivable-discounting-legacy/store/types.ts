import {
  IParticipantRFPSummary,
  IQuoteBase,
  IRFPSummariesResponse,
  ParticipantRFPStatus,
  ICargo,
  ITrade,
  RDStatus,
  IHistory
} from '@komgo/types'
import { Map } from 'immutable'
import { Action } from 'redux'
import { ImmutableMap, stringOrNull } from '../../../utils/types'
import { Counterparty } from '../../counterparties/store/types'
import { DocumentContextBase } from '../../trades/store/types'

// Only use REQUEST/SUCCESS/FAILURE at the end of action type names
export enum ReceivableDiscountingActionType {
  CREATE_REQUEST_FOR_PROPOSAL_REQUEST = '@@receivable-discounting/CREATE_RFP_REQUEST',
  CREATE_REQUEST_FOR_PROPOSAL_SUCCESS = '@@receivable-discounting/CREATE_RFP_SUCCESS',
  CREATE_REQUEST_FOR_PROPOSAL_FAILURE = '@@receivable-discounting/CREATE_RFP_FAILURE',

  FETCH_RFP_SUMMARY_REQUEST = '@@receivable-discounting/FETCH_RFP_SUMMARY_REQUEST',
  FETCH_RFP_SUMMARY_SUCCESS = '@@receivable-discounting/FETCH_RFP_SUMMARY_SUCCESS',
  FETCH_RFP_SUMMARY_FAILURE = '@@receivable-discounting/FETCH_RFP_SUMMARY_FAILURE',

  CREATE_QUOTE_REQUEST = '@@receivable-discounting/CREATE_QUOTE_REQUEST',
  CREATE_QUOTE_SUCCESS = '@@receivable-discounting/CREATE_QUOTE_SUCCESS',
  CREATE_QUOTE_FAILURE = '@@receivable-discounting/CREATE_QUOTE_FAILURE',

  FETCH_PARTICIPANT_RFP_SUMMARY_REQUEST = '@@receivable-discounting/FETCH_PARTICIPANT_RFP_SUMMARY_REQUEST',
  FETCH_PARTICIPANT_RFP_SUMMARY_SUCCESS = '@@receivable-discounting/FETCH_PARTICIPANT_RFP_SUMMARY_SUCCESS',
  FETCH_PARTICIPANT_RFP_SUMMARY_FAILURE = '@@receivable-discounting/FETCH_PARTICIPANT_RFP_SUMMARY_FAILURE',

  FETCH_QUOTE_REQUEST = '@@receivable-discounting/FETCH_QUOTE_REQUEST',
  FETCH_QUOTE_SUCCESS = '@@receivable-discounting/FETCH_QUOTE_SUCCESS',
  FETCH_QUOTE_FAILURE = '@@receivable-discounting/FETCH_QUOTE_FAILURE',

  SUBMIT_QUOTE_REQUEST = '@@receivable-discounting/SUBMIT_QUOTE_REQUEST',
  SUBMIT_QUOTE_SUCCESS = '@@receivable-discounting/SUBMIT_QUOTE_SUCCESS',
  SUBMIT_QUOTE_FAILURE = '@@receivable-discounting/SUBMIT_QUOTE_FAILURE',

  REJECT_RFP_REQUEST = '@@receivable-discounting/REJECT_RFP_REQUEST',
  REJECT_RFP_SUCCESS = '@@receivable-discounting/REJECT_RFP_SUCCESS',
  REJECT_RFP_FAILURE = '@@receivable-discounting/REJECT_RFP_FAILURE',

  ACCEPT_QUOTE_REQUEST = '@@receivable-discounting/ACCEPT_QUOTE_REQUEST',
  ACCEPT_QUOTE_SUCCESS = '@@receivable-discounting/ACCEPT_QUOTE_SUCCESS',
  ACCEPT_QUOTE_FAILURE = '@@receivable-discounting/ACCEPT_QUOTE_FAILURE',

  FETCH_TRADE_HISTORY_REQUEST = '@@receivable-discounting/FETCH_TRADE_HISTORY_REQUEST',
  FETCH_TRADE_HISTORY_SUCCESS = '@@receivable-discounting/FETCH_TRADE_HISTORY_SUCCESS',
  FETCH_TRADE_HISTORY_FAILURE = '@@receivable-discounting/FETCH_TRADE_HISTORY_FAILURE'
}

export interface IMemberMarketSelectionItem {
  counterparty: Counterparty
  location: string
  appetite: string
  availability: string
  creditLimit: string
  riskFee: string
  margin: string
  maxTenor: string
}

export interface IReceivableDiscountingDashboardTrader {
  tradeTechnicalId: string
  rdId?: string
  tradeDate: string | number | Date
  tradeId: string
  counterparty: string
  bank: string
  commodity: string
  invoiceAmount?: string
  invoiceType?: string
  currency: string
  status: string
  rdStatus?: RDStatus
}

export interface IReceivableDiscountingDashboardBank {
  tradeId: string
  seller: string
  buyer: string
  paymentTerms: string
  rd: {
    status: RDStatus
    staticId: string
  }
  invoiceAmount: string
  invoiceType: string
  currency: string
  discountingDate: string | number | Date
  discountingDateType: string
  status: string
  requestDate: string
}

export interface IRFPRequestSummary {
  companyStaticId: string
  status: ParticipantRFPStatus
}

export interface ISubmitQuoteFormDetails extends IQuoteBase {
  comment?: string
  rdId: string
}

// Move to @komgo/types
export interface IRFPReply {
  rdId: string
  comment?: string
}

export interface IBankDeclineRFPFormDetails {
  comment?: string
}

export interface IFetchMultipleReceivableDiscountFilter {
  tradeSourceIds: string[]
}

export interface ICreateRequestForProposal {
  rdId: string
  participantStaticIds: string[]
}

export interface ITradeSnapshot {
  source: string
  sourceId: string
  trade: ITrade
  movements: ICargo[]
  createdAt?: string
  updatedAt?: string
}

// Move to @komgo/types
export interface IQuoteSubmission extends IRFPReply {
  quoteId: string
}

// Move to @komgo/types
export interface IQuoteAcceptSubmission extends IQuoteSubmission {
  participantStaticId: string
}

export interface RDDocumentContext extends DocumentContextBase {
  rdId: string
}

export interface CreateRequestForProposalError extends Action {
  type: ReceivableDiscountingActionType.CREATE_REQUEST_FOR_PROPOSAL_FAILURE
  payload: stringOrNull
}

export interface ReceivableDiscountingFetchedQuotesAction extends Action {
  type: ReceivableDiscountingActionType.FETCH_RFP_SUMMARY_SUCCESS
  payload: IRFPSummariesResponse
  rdId: string
}

export interface ReceivableDiscountingFetchedTradeHistoryAction extends Action {
  type: ReceivableDiscountingActionType.FETCH_TRADE_HISTORY_SUCCESS
  payload: IHistory<ITradeSnapshot>
  sourceId: string
}

export interface ReceivableDiscountingStateProperties {
  rfpSummariesByRdId: Map<string, IParticipantRFPSummary[]>
  rfpSummariesByRdIdByParticipantStaticId: Map<string, Map<string, IParticipantRFPSummary>>
  tradeSnapshotHistoryById: Map<string, IHistory<ITradeSnapshot>>
  error: stringOrNull
}

export type ReceivableDiscountingState = ImmutableMap<ReceivableDiscountingStateProperties>

export type ReceivableDiscountingAction =
  | ReceivableDiscountingFetchedQuotesAction
  | CreateRequestForProposalError
  | ReceivableDiscountingFetchedTradeHistoryAction
