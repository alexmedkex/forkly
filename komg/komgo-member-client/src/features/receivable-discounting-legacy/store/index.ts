import * as actions from './actions'
import reducer from './reducer'

export { reducer as ReceivableDiscountingReducer, actions as ReceivableDiscountingActions }
