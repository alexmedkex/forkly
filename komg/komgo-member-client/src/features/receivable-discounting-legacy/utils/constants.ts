import {
  Currency,
  InvoiceType,
  QUOTE_SCHEMA,
  RECEIVABLES_DISCOUNTING_SCHEMA,
  FinancialInstrument,
  IQuoteBase
} from '@komgo/types'
import { RequestType } from '@komgo/types/dist/receivables-discounting/models/RequestType'
import { DiscountingType } from '@komgo/types/dist/receivables-discounting/models/DiscountingType'
import { SchemaUtils } from './SchemaUtils'
import { IRadioButtonWithDescription } from '../../../components/form/radio-button-group-with-descriptions/RadioButtonGroupWithDescriptions'
import { displayRequestType, displayMaturity } from './displaySelectors'
import { Strings } from '../resources/strings'
import { QuoteInputFieldLogic } from '../presentation/QuoteInputFieldLogic'

// TODO: move to komgo/types
export enum ReceivablesDiscountingRole {
  Trader = 'TRADER',
  Bank = 'BANK'
}

export enum DiscountingDateType {
  Expected = 'EXPECTED',
  Discounted = 'DISCOUNTED'
}

export enum ReceivableDiscountingViewPanels {
  TradeSummary = 'TRADE_SUMMARY',
  ReceivableDiscountingData = 'RECEIVABLE_DISCOUNTING DATA',
  SubmittedQuote = 'SUBMITTED_QUOTE',
  AcceptedQuote = 'ACCEPTED_QUOTE',
  Documents = 'DOCUMENTS'
}

export enum ApplyForDiscountingPanels {
  TradeSummary = 'TRADE_SUMMARY',
  ApplyForDiscountingData = 'APPLY_FOR_DISCOUNTING_DATA'
}

export enum ReceivableDiscountingModal {
  Submit = 'SUBMIT',
  Decline = 'DECLINE'
}

export const initialFinancialInstrumentInfo = {
  financialInstrument: FinancialInstrument.SBLC,
  financialInstrumentIssuerName: '',
  financialInstrumentIfOther: ''
}

export const RD_DEFAULT_VERSION = 2
// Needs to be type `any` based on some validation between fields of type number.
export const initialApplyForDiscountingData = (requestType?: RequestType, discountingType?: DiscountingType): any => {
  if (!requestType) {
    // return default object if no requestType is selected
    return { supportingInstruments: [] }
  } else {
    return {
      version: RD_DEFAULT_VERSION,
      tradeReference: { sourceId: '', sellerEtrmId: '', source: '' },
      requestType,
      discountingType: discountingType ? discountingType : DiscountingType.WithoutRecourse,
      invoiceAmount: 0,
      currency: Currency.USD,
      invoiceType: InvoiceType.Indicative,
      supportingInstruments: [],
      advancedRate: undefined,
      dateOfPerformance: undefined,
      discountingDate: requestType === RequestType.Discount ? '' : undefined,
      numberOfDaysDiscounting: requestType === RequestType.Discount ? ('' as any) : undefined,
      riskCoverDate:
        requestType !== RequestType.Discount || discountingType === DiscountingType.Blended ? '' : undefined,
      numberOfDaysRiskCover:
        requestType !== RequestType.Discount || discountingType === DiscountingType.Blended ? '' : undefined,
      guarantor: '',
      financialInstrumentInfo: initialFinancialInstrumentInfo,
      comment: undefined
    }
  }
}

// Needs to be type `any` based on some validation between fields of type number.
export const initialSubmitQuoteData = (
  rdCurrency: Currency,
  requestType: RequestType,
  discountingType?: DiscountingType
): any => {
  const quoteDefaultValue = (fieldName: string, defaultValue?: any) =>
    SchemaUtils.findDefaultValue(QUOTE_SCHEMA, fieldName, requestType, discountingType) || defaultValue

  const baseFields = {
    advanceRate: quoteDefaultValue('advanceRate', ''),
    pricingType: quoteDefaultValue('pricingType', ''),
    comment: quoteDefaultValue('comment', ''),
    pricingAllIn: quoteDefaultValue('pricingAllIn', ''),
    pricingRiskFee: quoteDefaultValue('pricingRiskFee', ''),
    pricingMargin: quoteDefaultValue('pricingMargin', ''),
    pricingFlatFeeAmount: quoteDefaultValue('pricingFlatFeeAmount', {
      amount: 0,
      currency: rdCurrency
    })
  }

  const riskCoverFields = {
    numberOfDaysRiskCover: quoteDefaultValue('numberOfDaysRiskCover', '')
  }

  const discountingFields = {
    numberOfDaysDiscounting: quoteDefaultValue('numberOfDaysDiscounting', '')
  }

  const interestFields = {
    interestType: quoteDefaultValue('interestType', ''),
    addOnValue: quoteDefaultValue('addOnValue', ''),
    indicativeCof: quoteDefaultValue('indicativeCof', ''),
    feeCalculationType: quoteDefaultValue('feeCalculationType', ''),
    liborType: quoteDefaultValue('liborType', ''),
    daysUntilMaturity: quoteDefaultValue('daysUntilMaturity', '')
  }

  let initialValues = baseFields

  if (QuoteInputFieldLogic.shouldShowDiscountingFields(requestType)) {
    initialValues = { ...initialValues, ...discountingFields }
  }

  if (QuoteInputFieldLogic.shouldShowRiskCoverFields(requestType, discountingType)) {
    initialValues = { ...initialValues, ...riskCoverFields }
  }

  if (QuoteInputFieldLogic.shouldShowInterestTypeFields(requestType)) {
    initialValues = { ...initialValues, ...interestFields }
  }

  return initialValues
}

export const initialAcceptQuoteData = (
  quote: IQuoteBase,
  rdCurrency: Currency,
  requestType: RequestType,
  discountingType?: DiscountingType
): any => {
  const quoteDefaultValue = (fieldName: string, defaultValue?: any) => quote[fieldName] || defaultValue

  const baseFields = {
    advanceRate: quoteDefaultValue('advanceRate', ''),
    pricingType: quoteDefaultValue('pricingType', ''),
    comment: quoteDefaultValue('comment', ''),
    pricingAllIn: quoteDefaultValue('pricingAllIn', ''),
    pricingRiskFee: quoteDefaultValue('pricingRiskFee', ''),
    pricingMargin: quoteDefaultValue('pricingMargin', ''),
    pricingFlatFeeAmount: quoteDefaultValue('pricingFlatFeeAmount', {
      amount: 0,
      currency: rdCurrency
    })
  }

  const riskCoverFields = {
    numberOfDaysRiskCover: quoteDefaultValue('numberOfDaysRiskCover', '')
  }

  const discountingFields = {
    numberOfDaysDiscounting: quoteDefaultValue('numberOfDaysDiscounting', '')
  }

  const interestFields = {
    interestType: quoteDefaultValue('interestType', ''),
    addOnValue: quoteDefaultValue('addOnValue', ''),
    indicativeCof: quoteDefaultValue('indicativeCof', ''),
    feeCalculationType: quoteDefaultValue('feeCalculationType', ''),
    liborType: quoteDefaultValue('liborType', ''),
    daysUntilMaturity: quoteDefaultValue('daysUntilMaturity', '')
  }

  let initialValues = baseFields

  if (QuoteInputFieldLogic.shouldShowDiscountingFields(requestType)) {
    initialValues = { ...initialValues, ...discountingFields }
  }

  if (QuoteInputFieldLogic.shouldShowRiskCoverFields(requestType, discountingType)) {
    initialValues = { ...initialValues, ...riskCoverFields }
  }

  if (QuoteInputFieldLogic.shouldShowInterestTypeFields(requestType)) {
    initialValues = { ...initialValues, ...interestFields }
  }

  return initialValues
}

export const initialBankDeclineRFPData: any = {
  comment: ''
}

export interface IMemberMarketSelection {
  provider: string
}

export const rdDiscountingSchema = RECEIVABLES_DISCOUNTING_SCHEMA
export const rdQuoteSchema = QUOTE_SCHEMA

export const ONE_MONTH_MATURITY_DAYS = 30
export const authorizedMaturityValues = [
  1,
  7,
  ONE_MONTH_MATURITY_DAYS,
  ONE_MONTH_MATURITY_DAYS * 2,
  ONE_MONTH_MATURITY_DAYS * 3,
  ONE_MONTH_MATURITY_DAYS * 6,
  ONE_MONTH_MATURITY_DAYS * 12
]
export const maturityLabels = authorizedMaturityValues
  .map(v => ({ [`${v}`]: displayMaturity(v) }))
  .reduce((p, c) => ({ ...p, ...c }), {})

export const applyForDiscountingRadioButtonOptions: IRadioButtonWithDescription[] = [
  {
    value: RequestType.RiskCover,
    label: displayRequestType(RequestType.RiskCover),
    description: Strings.RiskCoverDescription
  },
  {
    value: RequestType.RiskCoverDiscounting,
    label: displayRequestType(RequestType.RiskCoverDiscounting),
    description: Strings.RiskCoverReceivableDiscountingDescription
  },
  {
    value: RequestType.Discount,
    label: displayRequestType(RequestType.Discount),
    description: Strings.ReceivableDiscountingDescription
  }
]

export enum ReturnContext {
  RDViewRequest = 'rdViewRequest',
  TradeDashboard = 'tradeDashboard'
}
