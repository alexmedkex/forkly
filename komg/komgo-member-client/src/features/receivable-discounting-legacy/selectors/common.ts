import { ITradeEnriched } from '../../trades/store/types'
import { TradingRole } from '../../trades/constants'

export const selectSellerTrades = (trades: ITradeEnriched[]): ITradeEnriched[] => {
  return trades.filter(trade => trade.tradingRole === TradingRole.SELLER)
}
