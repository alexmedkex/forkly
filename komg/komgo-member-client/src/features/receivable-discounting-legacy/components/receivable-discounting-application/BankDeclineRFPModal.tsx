import { IReceivablesDiscountingInfo } from '@komgo/types'
import { connect, Field, Formik, FormikContext, FormikProps } from 'formik'
import * as React from 'react'
import { Button, Form, Header, Modal } from 'semantic-ui-react'
import Text from '../../../../components/text'
import { stringOrNull } from '../../../../utils/types'
import { TextAreaController } from '../../../letter-of-credit-legacy/components/InputControllers'
import { findFieldFromSchema } from '../../../../store/common/selectors/displaySelectors'
import { IBankDeclineRFPFormDetails, IRFPReply } from '../../store/types'
import { initialBankDeclineRFPData, rdQuoteSchema, ReceivableDiscountingModal } from '../../utils/constants'
import SubmitStatus from './SubmitStatus'
import { displaySimpleRequestType } from '../../utils/displaySelectors'

export interface BankDeclineRFPModalOwnProps {
  visible: boolean
  tradeId: string
  discountingRequest?: IReceivablesDiscountingInfo
  sellerName?: string
  buyerName?: string
  rdId: string
  rdError: stringOrNull
  bankDeclineRFPLoader: boolean
  toggleVisible(modal: string): void
  bankDeclineRFP(values: IRFPReply): void
}

interface BankDeclineRFPModalState {
  values: IBankDeclineRFPFormDetails
  openConfirm?: boolean
}

export type BankDeclineRFPModalProps = BankDeclineRFPModalOwnProps & {
  formik: FormikContext<IBankDeclineRFPFormDetails>
}

class BankDeclineRFPModal extends React.Component<BankDeclineRFPModalProps, BankDeclineRFPModalState> {
  constructor(props: BankDeclineRFPModalProps) {
    super(props)
    this.state = {
      openConfirm: false,
      values: initialBankDeclineRFPData
    }
  }

  handleSubmit = (values: IBankDeclineRFPFormDetails, formik: FormikProps<IBankDeclineRFPFormDetails>) => {
    const { rdId, bankDeclineRFP } = this.props
    formik.setSubmitting(false)
    this.setState({
      openConfirm: true
    })

    const bankDeclineRFPData: IRFPReply = {
      comment: values.comment,
      rdId
    }

    bankDeclineRFP(bankDeclineRFPData)
  }

  handleCancelSubmit = () => {
    this.setState({
      openConfirm: false
    })
  }

  handleCloseModal = (formik: FormikProps<IBankDeclineRFPFormDetails>) => {
    formik.resetForm()
    this.props.toggleVisible(ReceivableDiscountingModal.Decline)
  }

  render() {
    const { tradeId, sellerName, buyerName, rdError, bankDeclineRFPLoader, visible, discountingRequest } = this.props

    return (
      <Formik
        initialValues={initialBankDeclineRFPData}
        onSubmit={this.handleSubmit}
        render={(formik: FormikProps<IBankDeclineRFPFormDetails>) => {
          const { handleSubmit } = formik

          return (
            <React.Fragment>
              <Modal
                data-test-id="bank-decline-rfp-modal"
                open={visible}
                centered={true}
                style={{ top: 'unset', width: '680px' }}
              >
                <Modal.Content>
                  <Header as="h1" content="Decline discounting request" />
                  <p>
                    You are declining a{' '}
                    {discountingRequest
                      ? `${displaySimpleRequestType(discountingRequest.rd.requestType).toLocaleLowerCase()} request`
                      : 'request'}{' '}
                    for <Text bold={true}>Trade ID {tradeId}</Text> from {sellerName} for a quote on{' '}
                    <Text bold={true}>{buyerName}</Text>.
                  </p>
                  <Form onSubmit={handleSubmit} id="bank-decline-rfp-form">
                    <label style={{ fontWeight: 'bold' }}>
                      {findFieldFromSchema('title', 'comment', rdQuoteSchema)}
                    </label>
                    <Field component={TextAreaController} name="comment" data-test-id="comment" />
                  </Form>
                </Modal.Content>
                <Modal.Actions>
                  <Button
                    className="ui button"
                    role="button"
                    onClick={() => this.handleCloseModal(formik)}
                    data-test-id="bank-decline-rfp-cancel-btn"
                  >
                    Cancel
                  </Button>
                  <Button
                    primary={true}
                    content="Decline request"
                    form="bank-decline-rfp-form"
                    type="submit"
                    data-test-id="bank-decline-rfp-submit-btn"
                  />
                </Modal.Actions>
                <SubmitStatus
                  title="Decline Request"
                  actionText="Declining request"
                  isSubmitting={bankDeclineRFPLoader}
                  open={this.state.openConfirm}
                  cancelSubmit={this.handleCancelSubmit}
                  error={rdError}
                />
              </Modal>
            </React.Fragment>
          )
        }}
      />
    )
  }
}

export default connect<BankDeclineRFPModalOwnProps, IBankDeclineRFPFormDetails>(BankDeclineRFPModal)
