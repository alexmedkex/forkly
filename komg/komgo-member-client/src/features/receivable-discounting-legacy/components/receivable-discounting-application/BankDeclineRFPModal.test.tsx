import * as React from 'react'
import { mount } from 'enzyme'
import * as renderer from 'react-test-renderer'
import BankDeclineRFPModal, { BankDeclineRFPModalOwnProps } from './BankDeclineRFPModal'
import { initialBankDeclineRFPData } from '../../utils/constants'
import { IBankDeclineRFPFormDetails } from '../../store/types'
import { FormikProvider } from 'formik'
import { Button } from 'semantic-ui-react'
import { fakeFormikContext } from '../../../../store/common/faker'
import { buildFakeReceivablesDiscountingInfo } from '@komgo/types'

// the magic https://github.com/Semantic-Org/Semantic-UI-React/issues/2454#issuecomment-373246622
jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children)

describe('<ReceivableDiscountingDeclineRequestModal />', () => {
  let wrapper
  let props: BankDeclineRFPModalOwnProps

  const formikContext = fakeFormikContext<IBankDeclineRFPFormDetails>(initialBankDeclineRFPData)

  beforeEach(() => {
    props = {
      visible: false,
      tradeId: 'sadfsdf',
      discountingRequest: buildFakeReceivablesDiscountingInfo(),
      sellerName: null,
      buyerName: null,
      rdId: 'An-RDID',
      rdError: null,
      bankDeclineRFPLoader: false,
      toggleVisible: jest.fn(),
      bankDeclineRFP: jest.fn()
    }

    wrapper = mount(
      <FormikProvider value={formikContext}>
        <BankDeclineRFPModal {...props} />
      </FormikProvider>
    )
  })

  it('should render as expected', () => {
    const jsonWrapper = renderer.create(<BankDeclineRFPModal {...props} />).toJSON()
    expect(jsonWrapper).toMatchSnapshot()
  })

  it('should toggle the modal when we click the cancel button', () => {
    expect(props.toggleVisible).toHaveBeenCalledTimes(0)

    wrapper
      .find({ 'data-test-id': 'bank-decline-rfp-cancel-btn' })
      .find(Button)
      .simulate('click')

    expect(props.toggleVisible).toHaveBeenCalledTimes(1)
  })
})
