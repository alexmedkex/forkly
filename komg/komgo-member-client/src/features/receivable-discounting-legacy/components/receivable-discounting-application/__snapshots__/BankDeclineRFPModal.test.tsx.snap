// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReceivableDiscountingDeclineRequestModal /> should render as expected 1`] = `
.c0 {
  display: inline-block;
  font-size: 14px;
  font-weight: bold;
  color: #000000;
  margin: 0;
}

.c2 {
  display: inline-block;
  font-size: 14px;
  font-weight: normal;
  color: #142435;
  margin: 0;
}

.c1 {
  margin-top: 0;
  position: relative;
  background-color: rgba(255,255,255,0.85);
}

<div
  className="ui page modals dimmer transition visible active"
>
  <div
    className="ui modal transition visible active"
    data-test-id="bank-decline-rfp-modal"
    style={
      Object {
        "marginTop": undefined,
        "top": "unset",
        "width": "680px",
      }
    }
  >
    <div
      className="content"
    >
      <h1
        className="ui header"
      >
        Decline discounting request
      </h1>
      <p>
        You are declining a
         
        receivable discounting request
         
        for 
        <span
          className="c0"
        >
          Trade ID 
          sadfsdf
        </span>
         from 
         for a quote on
         
        <span
          className="c0"
        />
        .
      </p>
      <form
        className="ui form"
        id="bank-decline-rfp-form"
        onSubmit={[Function]}
      >
        <label
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Comment (optional)
        </label>
        <textarea
          data-test-id="comment"
          name="comment"
          onBlur={[Function]}
          onChange={[Function]}
          onInput={[Function]}
          rows={3}
          style={
            Object {
              "background": undefined,
              "border": undefined,
              "resize": "",
            }
          }
          value=""
        />
      </form>
    </div>
    <div
      className="actions"
    >
      <button
        className="ui button ui button"
        data-test-id="bank-decline-rfp-cancel-btn"
        onClick={[Function]}
        role="button"
      >
        Cancel
      </button>
      <button
        className="ui primary button"
        data-test-id="bank-decline-rfp-submit-btn"
        form="bank-decline-rfp-form"
        onClick={[Function]}
        role="button"
        type="submit"
      >
        Decline request
      </button>
    </div>
    <div
      className="ui page modals dimmer transition visible active"
    >
      <div
        className="ui tiny modal transition visible active"
        style={
          Object {
            "marginTop": undefined,
          }
        }
      >
        <div
          className="header"
        >
          Declining request
        </div>
        <div
          className="content"
        >
          <div
            className="c1"
            data-test-id="loadingTransition"
          >
            <img
              className="ui mini centered image"
              src="/images/bars.svg"
            />
            <div
              className="ui header centered"
            >
              <span
                className="c2"
                color="#142435"
              >
                Declining request
              </span>
            </div>
          </div>
        </div>
        <div
          className="actions"
        >
          <button
            className="ui button"
            data-test-id="cancel-button"
            onClick={[Function]}
            role="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
